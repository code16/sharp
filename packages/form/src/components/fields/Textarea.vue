<template>
    <textarea 
        class="SharpTextarea" 
        :rows="rows" 
        :placeholder="placeholder" 
        :disabled="readOnly"
        v-maxlength="maxLength" 
        @input="handleInput"
    >{{ value }}</textarea>
</template>

<script>
    import { maxlength } from 'sharp/directives';

    export default {
        name:'SharpTextarea',

        props: {
            value: {
                type: String,
            },
            placeholder: String,
            readOnly: Boolean,

            maxLength: Number,

            rows: Number,
        },
        data() {
            return { }
        },
        methods: {
            handleInput(e) {
                this.$emit('input', e.target.value);
            }
        },
        directives: {
            maxlength
        }
    }
</script>