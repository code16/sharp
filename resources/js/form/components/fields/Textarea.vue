<script setup lang="ts">
    import { FormTextareaFieldData } from "@/types";
    import { normalizeText } from "../../util/text";
    import { validateTextField } from "../../util/validation";
    import { Textarea } from "@/components/ui/textarea";
    import { FormFieldEmits, FormFieldProps } from "@/form/types";
    import FormFieldLayout from "@/form/components/FormFieldLayout.vue";

    const props = defineProps<FormFieldProps<FormTextareaFieldData>>();
    const emit = defineEmits<FormFieldEmits<FormTextareaFieldData>>();

    function onInput(textareaValue: string) {
        const value = normalizeText(textareaValue);
        const error = validateTextField(value, {
            maxlength: props.field.maxLength,
        });
        if(props.field.localized) {
            emit('input',
                typeof props.value === 'object'
                    ? { ...props.value, [props.locale]: value }
                    : { [props.locale]: value },
                { error }
            );
        } else {
            emit('input', value, { error });
        }
    }
</script>

<template>
    <FormFieldLayout v-bind="props" v-slot="{ id, ariaDescribedBy }">
        <Textarea
            :id="id"
            class="min-h-0"
            :model-value="field.localized && typeof value === 'object' ? value?.[locale] : (value as string)"
            :rows="field.rows"
            :placeholder="field.placeholder"
            :disabled="field.readOnly"
            :aria-describedby="ariaDescribedBy"
            @update:model-value="onInput"
        />
    </FormFieldLayout>
</template>
