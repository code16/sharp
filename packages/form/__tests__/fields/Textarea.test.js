import Vue from 'vue';
import Textarea from '../../src/components/fields/Textarea.vue';

describe('textarea-field',()=>{
    Vue.component('sharp-textarea', Textarea);

    beforeEach(()=>{
        document.body.innerHTML = `
            <div id="app">
                <sharp-textarea placeholder="Entrez du texte" :rows="30" read-only value="AAA" @input="inputEmitted($event)"></sharp-textarea>
            </div>
        `
    });

    test('can mount Textarea field', async () => {
        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    test('emit event on input', async () => {
        let inputEmitted = jest.fn();

        await createVm({
            methods: {
                inputEmitted
            }
        });
        let input = document.querySelector('textarea');
        input.dispatchEvent(new Event('input', { bubbles: true }));
        expect(inputEmitted.mock.calls.length).toBe(1);
    });

    test('emit correct value', async () => {
        let inputEmitted = jest.fn();

        await createVm({
            methods: {
                inputEmitted
            }
        });
        let input = document.querySelector('textarea');
        input.dispatchEvent(new Event('input', { bubbles: true }));
        expect(inputEmitted.mock.calls[0][0]).toBe('AAA');
    });
});


async function createVm(customOptions={}) {
    const vm = new Vue({
        el: '#app',
        mixins: [customOptions],
    });

    await Vue.nextTick();

    return vm.$children[0];
}