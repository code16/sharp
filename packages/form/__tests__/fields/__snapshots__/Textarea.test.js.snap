// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`textarea-field can mount Textarea field 1`] = `
<div id="app">
	<textarea rows="30" placeholder="Entrez du texte" disabled="disabled" class="SharpTextarea">AAA</textarea>
</div>
`;
